packages/bitcore-lib/lib/script/interpreter.js

```diff
@@ -70,26 +72,133 @@ Interpreter.prototype.verifyWitnessProgram = function(version, program, witness,
+        this.errstr = 'SCRIPT_ERR_WITNESS_PROGRAM_MISMATCH';
+        return false;
+      }
+      execdata.tapleafHashInit = true;
+      if ((control[0] & Interpreter.TAPROOT_LEAF_MASK) === Interpreter.TAPROOT_LEAF_TAPSCRIPT) {
+        // Tapscript (leaf version 0xc0)
+        let witnessSize;
+        {
+          const bw = new BufferWriter();
+          bw.writeVarintNum(witness.length);
+          for (let element of witness) {
+            bw.writeVarintNum(element.length);
+            bw.write(element);
+          }
+          witnessSize = bw.toBuffer().length;
+        }
+
+        try {
+          scriptPubKey = new Script(scriptPubKeyBuf);
+        } catch (err) {
+          // Note how this condition would not be reached if an unknown OP_SUCCESSx was found
+          this.errstr = 'SCRIPT_ERR_BAD_OPCODE';
+          return false;
+        }
+
+        execdata.validationWeightLeft = witnessSize + Script.VALIDATION_WEIGHT_OFFSET;
+        execdata.validationWeightLeftInit = true;
+        return this.executeWitnessScript(scriptPubKey, stack, Signature.Version.TAPSCRIPT, satoshis, flags, execdata);
+      }
+      if (flags & Interpreter.SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION) {
+        this.errstr = 'SCRIPT_ERR_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION';
+        return false;
+      }
+      return true;
+    }
   } else if ((flags & Interpreter.SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_WITNESS_PROGRAM)) {
     this.errstr = 'SCRIPT_ERR_DISCOURAGE_UPGRADABLE_WITNESS_PROGRAM';
     return false;
-  } else {
-    return true;
+  }
+  // Other version/size/p2sh combinations return true for future softfork compatibility
+  return true;
+};
+
+
+Interpreter.prototype.executeWitnessScript = function(scriptPubKey, stack, sigversion, satoshis, flags, execdata) {
+  if (sigversion === Signature.Version.TAPSCRIPT) {
+    for (let chunk of scriptPubKey.chunks) {
+      // New opcodes will be listed here. May use a different sigversion to modify existing opcodes.
+      if (Opcode.isOpSuccess(chunk.opcodenum)) {
+        if (flags & Interpreter.SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
+          this.errstr = 'SCRIPT_ERR_DISCOURAGE_OP_SUCCESS';
+          return false;
+        }
+        return true;
+      }
+    }
+
+    // Tapscript enforces initial stack size limits (altstack is empty here)
+    if (stack.length > Interpreter.MAX_STACK_SIZE) {
+      this.errstr = 'SCRIPT_ERR_STACK_SIZE';
+      return false;
+    }
+  }
+
+  // Disallow stack item size > MAX_SCRIPT_ELEMENT_SIZE in witness stack
+  if (stack.length && stack.some(elem => elem.length > Interpreter.MAX_SCRIPT_ELEMENT_SIZE)) {
+    this.errstr = 'SCRIPT_ERR_PUSH_SIZE';
+    return false;
   }
 
   this.initialize();
 
   this.set({
     script: scriptPubKey,
     stack: stack,
-    sigversion: Interpreter.SIGVERSION_WITNESS_V0,
+    sigversion: sigversion,
     satoshis: satoshis,
     flags: flags,
+    execdata: execdata
   });
 
   if (!this.evaluate()) {
```

Comment:
Proposed cleanup: add this.errstr for uniformity

Reply:
Similar cases below

Reply:
this.errstr is set inside this.evaluate()